xdem/spstats.py

```diff
@@ -9,19 +9,123 @@
+    list_df_1d = []
+    for i, var in enumerate(list_var):
+        df_stats_1d = pd.DataFrame()
+        # get statistics
+        for j, statistic in enumerate(statistics):
+            stats_binned_1d, bedges_1d = binned_statistic(var,dh,statistic=statistic,bins=list_var_bins[i],range=list_ranges)[:2]
+            # save in a dataframe
+            df_stats_1d[statistics_name[j]] = stats_binned_1d
+        # we need to get the middle of the bins from the edges, to get the same dimension length
+        df_stats_1d['bin_mid_var1'] = (bedges_1d[1:] + bedges_1d[:-1]) / 2
+        df_stats_1d['bin_low_var1'] = bedges_1d[:-1]
+        df_stats_1d['bin_upp_var1'] = bedges_1d[1:]
+        df_stats_1d['name_var1'] = list_var_names[i]
+        df_stats_1d['nd'] = 1
+
+        list_df_1d.append(df_stats_1d)
+
+    # get binned statistics in 2d: all possible 2d combinations
+    list_df_2d = []
+    if len(list_var)>1:
+        combs = list(itertools.combinations(list_var_names, 2))
+        for i, comb in enumerate(combs):
+            var1_name, var2_name = comb
+            # corresponding variables indexes
+            i1, i2 = list_var_names.index(var1_name), list_var_names.index(var2_name)
+            df_stats_2d = pd.DataFrame()
+            for j, statistic in enumerate(statistics):
+                stats_binned_2d, bedges_var1, bedges_var2 = binned_statistic_2d(list_var[i1],list_var[i2],dh,statistic=statistic
+                                                             ,bins=[list_var_bins[i1],list_var_bins[i2]]
+                                                             ,range=list_ranges)[:3]
+                # get statistics
+                df_stats_2d[statistics_name[j]] = stats_binned_2d.flatten()
+            bin_mid_var1 = (bedges_var1[1:] + bedges_var1[:-1]) / 2
+            bin_mid_var2 = (bedges_var2[1:] + bedges_var2[:-1]) / 2
+            df_stats_2d['bin_mid_var1'] = [b1 for b1 in bin_mid_var1 for b2 in bin_mid_var2]
+            df_stats_2d['bin_low_var1'] = [b1 for b1 in bedges_var1[:-1] for b2 in bin_mid_var2]
+            df_stats_2d['bin_upp_var1'] = [b1 for b1 in bedges_var1[1:] for b2 in bin_mid_var2]
+            df_stats_2d['name_var1'] = var1_name
+            df_stats_2d['bin_mid_var2'] = [b2 for b1 in bin_mid_var1 for b2 in bin_mid_var2]
+            df_stats_2d['bin_low_var2'] = [b2 for b1 in bin_mid_var1 for b2 in bedges_var2[:-1]]
+            df_stats_2d['bin_upp_var2'] = [b2 for b1 in bin_mid_var1 for b2 in bedges_var2[1:]]
+            df_stats_2d['name_var2'] = var2_name
+            df_stats_2d['nd'] = 2
+
+            list_df_2d.append(df_stats_2d)
+
+
+    # get binned statistics in nd, without redoing the same stats
+    df_stats_nd = pd.DataFrame()
+    if len(list_var)>2:
+        for j, statistic in enumerate(statistics):
+            stats_binned_2d, list_bedges = binned_statistic_dd(list_var,dh,statistic=statistic,bins=list_var_bins,range=list_ranges)[0:2]
```

Comment:
Can this nd case not be also applied to the 1D and 2D cases? This would make the code simpler...

Reply:
I could use the same function, yes, but I'd still need to do 3 separate loops to get the binning in all dimensions.
So nothing would change in the code except the function name of 1d, 2d and the format of their output variables, which would take a few more line to reshape in the right format.
 
The 1d, 2d functions were made so it would be easier to manipulate their outputs, so I think we shoud leave them as is